### PR TITLE
Stopper pattern

### PR DESCRIPTION
**Pull request by tiger5226**
stopper pattern, component unification

### CODE REVIEW

File: cluster/cluster.go [hunk 1]
```diff
@@ -29,8 +29,8 @@ type Cluster struct {
 	stop    *stopOnce.Stopper
 }
 
-// New returns a new Cluster instance that is not connected.
-func New(port int, seedAddr string) *Cluster {
+// NewCluster returns a new Cluster instance that is not connected.
+func NewCluster(port int, seedAddr string) *Cluster {
```
Comment (lyoshenka) [inline cluster/cluster.go]: there's no need to call this NewCluster unless there are multiple New things that can be created. thats the go convention - they prefer it look like `cluster.New()` when its called, over `cluster.NewCluster()`

File: dht/dht.go [hunk 2]
```diff
@@ -113,6 +113,7 @@ func New(config *Config) (*DHT, error) {
 
 // join makes current node join the dht network.
 func (dht *DHT) join() {
+	defer dht.stop.Done()
```
Comment (lyoshenka) [inline dht/dht.go]: i'd prefer `stop.Done()` be next to `stop.Add()`, so its clear what's happening. see example below

File: dht/dht.go [hunk 3]
```diff
@@ -157,8 +158,10 @@ func (dht *DHT) Start() error {
 		return err
 	}
 
-	dht.join()
-	dht.startReannouncer()
+	dht.stop.Add(1)
+	go dht.join()
```
Comment (lyoshenka) [inline dht/dht.go]: I'd rewrite these two lines like so:

```
dht.stop.Add(1)
go func(){
  dht.join()
  dht.stop.Done()
}()
```

this way Add() and Done() are near each other, and the join() function doesnt have to worry about how its called

File: dht/dht.go [hunk 4]
```diff
@@ -157,8 +158,10 @@ func (dht *DHT) Start() error {
...
+	dht.stop.Add(1)
+	go dht.startReannouncer()
```
Comment (lyoshenka) [inline dht/dht.go]: same here

File: dht/dht.go [hunk 5]
```diff
@@ -245,8 +255,6 @@ func (dht *DHT) Announce(hash Bitmap) error {
 }
 
 func (dht *DHT) startReannouncer() {
-	dht.stop.Add(1)
```
Comment (lyoshenka) [inline dht/dht.go]: i think this is fine. these can be together inside the function. they just can't be split across "inside" and "outside"

File: dht/dht.go [hunk 6]
```diff
@@ -151,14 +156,20 @@ func (dht *DHT) Start() error {
...
-
 	err = dht.node.Connect(conn)
 	if err != nil {
 		return err
 	}
-
-	dht.join()
-	dht.startReannouncer()
+	dht.stop.Add(1)
+	go func() {
```
Comment (lyoshenka) [inline dht/dht.go]: join has to finish before you can start the reannouncer. it can't be concurrent. there's no need to do these in goroutines

File: dht/dht.go [hunk 7]
```diff
@@ -135,6 +135,11 @@ func (dht *DHT) join() {
 
 	// now call iterativeFind on yourself
 	nf := newContactFinder(dht.node, dht.node.id, false)
+	// stop if dht is stopped
+	go func(finder *contactFinder) {
```
Comment (lyoshenka) [inline dht/dht.go]: you don't have to pass nf here. it will get nf from the outer scope. this is only necessary when using `go` or `defer` in a loop

Comment (lyoshenka) [reply dht/dht.go]: also this goroutine will stick around forever until dht.stop happens. it could be well after nf.Find() ran and finished

File: peer/server.go [hunk 8]
```diff
@@ -27,59 +28,69 @@ const (
...
 		} else {
+			s.stop.Add(1)
 			go s.handleConnection(conn)
 		}
 	}
 }
 
 func (s *Server) handleConnection(conn net.Conn) {
-	defer func(conn net.Conn) {
-		if err := conn.Close(); err != nil {
```
Comment (lyoshenka) [inline peer/server.go]: why did you delete this? the connection needs to be closed

File: reflector/prism.go [hunk 9]
```diff
@@ -33,21 +35,20 @@ func NewPrism(store store.BlobStore, clusterSeedAddr string) *Prism {
 	}
 }
 
-// Connect starts the components of the application.
-func (p *Prism) Connect() error {
-	err := p.dht.Start()
-	if err != nil {
+// Start starts the components of the application.
+func (p *Prism) Start() error {
+	if err := p.dht.Start(); err != nil {
```
Comment (lyoshenka) [inline reflector/prism.go]: could you write these on separate lines (and same in other places):
```
err := p.dht.Start()
if err != nil {
```
i think this makes it clear what's happening - first you start, then you check the error.

i see people do it the combined way sometimes, but i think they do that when the emphasis is on the if statement. here the emphasis is on the function call.

i could be wrong about this - happy to discuss and change my ways if you think that's the case.

File: reflector/server.go [hunk 10]
```diff
@@ -7,72 +7,81 @@ import (
...
 			go s.handleConn(conn)
 		}
 	}
 }
 
 func (s *Server) handleConn(conn net.Conn) {
+	defer s.stop.Done()
 	// TODO: connection should time out eventually
-	defer func(conn net.Conn) {
-		if err := conn.Close(); err != nil {
```
Comment (lyoshenka) [inline reflector/server.go]: conn needs to be closed

File: dht/dht.go (hunk 7 above)
Comment (lyoshenka) [reply dht/dht.go]: also, good catch that nf was not being closed correctly on shutdown

File: reflector/prism.go (hunk 9 above)
Comment (tiger5226) [reply reflector/prism.go]: I actually saw this recommended by someone as a good style for golang when I was doing code cleanup for Chainquery. The general rule of thumb is to keep these scoped inside the `if` statement if:

1) there is a single return 
2) and it is an error. 
3) and there are no complex expressions causing it to lose readability. 

There are also times when you don't want it scoped too which is in the style guide as the exception. 

Personally, I think it looks cleaner and I see a lot of other golang projects do it too. Maybe its not a style guide thing https://github.com/golang/go/wiki/CodeReviewComments#handle-errors  but it does leave it a bit vague. Basically it says if there is a variable init than you should break it out. I don' think I can say the opposite statement is endorsed or preferred, but I think a lot of people assume it is. 

I think you are the leader/owner here, so if I did not convince, happy to make sure I follow a style you prefer. Let me know.

File: reflector/server.go (hunk 10 above)
Comment (tiger5226) [reply reflector/server.go]: but this is done https://github.com/lbryio/reflector.go/blob/master/reflector/server.go#L59. I just encapsulated it. It looked to me like it was intentionally the last thing before shutdown with the defer. I could be wrong though.

File: peer/server.go (hunk 8 above)
Comment (tiger5226) [reply peer/server.go]: but this is done https://github.com/lbryio/reflector.go/blob/master/peer/server.go#L69-L75. I just encapsulated it. It looked to me like it was intentionally the last thing before shutdown with the defer. I could be wrong though.

File: dht/dht.go (hunk 7 above)
Comment (tiger5226) [reply dht/dht.go]: yeah, this was actually really annoying. I thought I had broken everything because the shutdown just wasn't working even though I had done everything I thought was right. Eventually, I stepped away for a minute without killing the terminal, and came back to it closed. So I turned on verbose mode, and let it sit, then traced the text in the code that was outputted. It lead me here. Thank god, because I almost gave up and was gonna bring you in since you know the code better than me.

File: dht/dht.go (hunk 5 above)
Comment (tiger5226) [reply dht/dht.go]: Actually I did google that before I did it. As with the earlier stuff, I was concerned about race conditions but because it is inside the scope of a function it should never cause a race condition.

File: peer/server.go (hunk 8 above)
Comment (lyoshenka) [reply peer/server.go]: closing the listener and closing the connection are two separate things. they both need to happen.
